### PR TITLE
Native triggers and reentrancy

### PR DESCRIPTION
**Pull request by pdonahue-ventana**
My proposal to address #653.

### CODE REVIEW

**Comment by timsifive** [comment]
This reads like a good clarification, although it now requires one of the two solutions be implemented. Was that your intention? It's not strictly backwards compatible if we require one of the two solutions. Did I miss something and are the proposed solutions actually different?

**Comment by pdonahue-ventana** [comment]
It was my intention to add that as a requirement if it's possible for any of the enumerated situations to exist.  The 0.13 spec said that this was a remote corner case but I think that's not really true, especially when delegation comes into play.  Without this, tmexttrigger with action=0 is useless and etrigger/itrigger with action=0 are definitely hampered.  We can, of course, debate these things.  If it remains optional in this spec (with just a "debuggers need to take care not to cause infinite loops of traps" warning) then I think that platform specs should make it mandatory.

I don't quite understand your last question.  The two solutions are definitely different from each other since the first doesn't require tcontrol.  The first solution is the same as 0.13 but with the addition of masking at lower privilege modes when appropriate.  The second solution is the same as 0.13 but with the addition of hard-wiring medeleg to prevent having the problem at lower privilege modes (which happens naturally for M or M+U implementations but could also work for M+S+U).

**Comment by timsifive** [comment]
The change is OK by me, but it is definitely not a backwards-compatible change. We've been saying that the spec is stable and won't see functional changes between now and ratification.

So... We could change it anyway. We'd have to invent some extra process to discourage us from doing that in the future. We could start a 1.1 fork and put this change in there.

What do you think?

**Comment by ernie-sifive** [comment]
I would argue that once the spec is "Stable" no further functional changes can be made.  Implementations of v1.0 already exist.  I think this change should be put into v1.1.

**Comment by pdonahue-ventana** [comment]
I knew that would be controversial and I certainly understand why.

How about this?
- "Harts that support triggers with \FcsrMcontrolSixAction=0 **should** implement one of the following two solutions"
- change "should" to "must" in 1.1

I think that it's important to talk about the problem and at least provide guidance.  That's what the 0.13 spec did, though I think that it didn't go far enough since it didn't contemplate delegation (which is what this change addresses).

File: trigger.tex [hunk 1]
```diff
@@ -142,33 +142,84 @@ \section{Priority}
...
+
+In these cases such a trigger may cause a breakpoint exception while already
+in a trap handler. This might leave the hart unable to resume normal execution
+because state such as \Rmcause and \Rmepc would be overwritten.
+\end{commentary}
+
+\begin{steps}{Harts that support triggers with \FcsrMcontrolSixAction=0
+should implement one of the following two solutions to solve the problem of
+reentrancy:}
+\item The hardware must prevent triggers with \FcsrMcontrolSixAction=0 from
```
Comment (timsifive) [inline trigger.tex:191]: Must in these two items is confusing. I'd leave the line above (which now says should) and then describe the behavior without must/should. E.g. "The hardware prevents triggers ..."

**Review by timsifive — CHANGES_REQUESTED** [review]
Overall this looks good. Just a language comment below.

File: trigger.tex (hunk 1 above)
Comment (pdonahue-ventana) [reply trigger.tex:191]: I obviously didn't reread the whole thing when making that must->should change.  I'll reread the whole thing and make changes along the lines of your suggestion.

**Review by timsifive — APPROVED** [review]

**Pull request merged**